### ISSUE TITLE
wd_capabilities.au3 - TODO better error handling

### ISSUE THREAD

**Issue by mlipok** [Danp2/au3WebDriver#105]
as a project

**Comment by Danp2** [comment]
```
Func _WD_CapabilitiesAdd($key, $value1 = '', $value2 = '')
	If $value1 = Default Then $value1 = 'default'
	If $value2 = Default Then $value2 = 'default'
````

What purpose do these If statements serve? 

Also, would it be better to have this instead for the Func declaration?

`Func _WD_CapabilitiesAdd($key, $value1 = Default, $value2 = Default)`

**Comment by Danp2** [comment]
`	If StringInStr('alwaysMatch|firstMatch', $key) Then`

This can be problematic because it also matches a partial string (`always`, `first`, `rstmat`, etc).

**Comment by mlipok** [comment]
> What purpose do these If statements serve?

```
_WD_CapabilitiesAdd2('args', '--profile-directory', Default)
_WD_CapabilitiesAdd2('args', '--profile-directory', "default")
```
this two lines do the same thing

If I would not use:
```
If $value1 = Default Then $value1 = 'default'
If $value2 = Default Then $value2 = 'default'
```

then this code:
```
_WD_CapabilitiesAdd2('args', '--profile-directory', Default)
```
will be translated to the following:
```
_WD_CapabilitiesAdd2('args', '--profile-directory', "Default")
```
which will fail.

**Comment by mlipok** [comment]
> Also, would it be better to have this instead for the Func declaration?
> 
> `Func _WD_CapabilitiesAdd($key, $value1 = Default, $value2 = Default)`

As so far, my answer is: 
no, because of my answer given above.

**Comment by mlipok** [comment]
> ` If StringInStr('alwaysMatch|firstMatch', $key) Then`
> 
> This can be problematic because it also matches a partial string (`always`, `first`, `rstmat`, etc).

in my WIP code.... I already use:
```
If StringInStr($key, 'alwaysMatch|firstMatch') Then
```
but we can even switch to StringRegExp or use:
```
If $key = 'alwaysMatch' Or  $key = 'firstMatch' Then
```

**Comment by Danp2** [comment]
> in my WIP code.... I already use:
>
> `If StringInStr($key, 'alwaysMatch|firstMatch') Then`

I don't see how that will even work. 😦

**Comment by mlipok** [comment]
> > in my WIP code.... I already use:
> > `If StringInStr($key, 'alwaysMatch|firstMatch') Then`
> 
> I don't see how that will even work. 😦

You are right, it was my mistake.

As I sad this following will be used.

> ```
> If $key = 'alwaysMatch' Or  $key = 'firstMatch' Then
> ```
> The latter is the simplest, fully functional, so I think we will use it.

**Comment by mlipok** [comment]
Please check this conceptual validation:

```
Func _WD_CapabilitiesAdd2($key, $value1 = '', $value2 = '', $CapType_SetByUser = Null)
	Local Static $s_MatchType = '', $s_SpecificOptions_KeyName = '', $s_DeclaredBrowsers = ''
	If StringInStr($key, 'alwaysMatch|firstMatch') Then
	If $key = 'alwaysMatch' Or  $key = 'firstMatch' Then
		Local $s_BrowserType = $value1

		; validate if you can change $s_MatchType
		If $key = 'alwaysMatch' And $_WD_CAPS__OBJECT <> '' Then Return SetError(1) ; you can use 'alwaysMatch' only at the begining - because of simplicfity of this UDF
		If $key = 'firstMatch' And $s_BrowserType = '' Then Return SetError(2) ; you should not use 'firstMatch' without specifing $s_BrowserType - because of simplicfity of this UDF
		If StringInStr($s_DeclaredBrowsers, $s_BrowserType) Then Return SetError(3) ; in firstMatch you should not declare again $s_BrowserType which was already used - because of simplicfity of this UDF


```


ps. 
the 2 number in the function name `_WD_CapabilitiesAdd2()` is used intentinally and is temporary in my testing UDF

**Comment by Danp2** [comment]
> The latter is the simplest, fully functional, so I think we will use it.

You could do it like this if you wanted to continue using StringInStr --

`If StringInStr('|alwaysMatch|firstMatch|', '|' & $key & '|') Then`

😃

**Comment by Danp2** [comment]
> Func _WD_CapabilitiesAdd2($key, $value1 = '', $value2 = '', $CapType_SetByUser = Null)

FWIW, I believe all optional parameters in the other UDF functions default to the keyword Default. 

>	If StringInStr($key, 'alwaysMatch|firstMatch') Then
>	If $key = 'alwaysMatch' Or  $key = 'firstMatch' Then

Why is the first If statement still present? Can't that be eliminated?

**Comment by mlipok** [comment]
> Can't that be eliminated?

Must be, and now it is ;)

**Comment by mlipok** [comment]
> FWIW, I believe all optional parameters in the other UDF functions default to the keyword Default.

there is only one other function with default value set for paramter:
`Func __WD_CapabilitiesInitialize($s_MatchType, $s_BrowserType = '')`

This function will not be needed soon.

**Comment by mlipok** [comment]
> Please check this conceptual validation:

especially all `because of simplicfity of this UDF`

**Comment by mlipok** [comment]
> Have you considered using Switch...Case...EndSwitch instead of the nested If...EndIf statements?

will try to switch ;)

**Comment by mlipok** [comment]
Please assign this to me.

**Comment by Danp2** [comment]
Needs to use __WD_Error throughout

**Comment by mlipok** [comment]
> Needs to use __WD_Error throughout

I hope now it fits yours requirments.
I mean after all changes in:
https://github.com/Danp2/au3WebDriver/pull/263

**Comment by Danp2** [comment]
> > ` If StringInStr('alwaysMatch|firstMatch', $key) Then`
> > This can be problematic because it also matches a partial string (`always`, `first`, `rstmat`, etc).
> 
> in my WIP code.... I already use:
> 
> ```
> If StringInStr($key, 'alwaysMatch|firstMatch') Then
> ```
> 
> but we can even switch to StringRegExp or use:
> 
> ```
> If $key = 'alwaysMatch' Or  $key = 'firstMatch' Then
> ```
> 
> The latter is the simplest, fully functional, so I think we will use it.

Fixed in 99ca6f65dd4dd302555bbc9d38e18f6ef830cc29

**Comment by Danp2** [comment]
> > Also, would it be better to have this instead for the Func declaration?
> > `Func _WD_CapabilitiesAdd($key, $value1 = Default, $value2 = Default)`
> 
> As so far, my answer is: no, because ... look at my previous answer.

When you have some time, I would like to discuss this further.

**Comment by mlipok** [comment]
error handling should be checked

_Originally posted by @mlipok in https://github.com/Danp2/au3WebDriver/pull/263#discussion_r830701133_

**Comment by mlipok** [comment]
> > Have you considered using Switch...Case...EndSwitch instead of the nested If...EndIf statements?
> 
> will try to switch ;)

I do not see where I could use switch instead If...EndIF

**Comment by mlipok** [comment]
> ` If StringInStr('alwaysMatch|firstMatch', $key) Then`
> 
> This can be problematic because it also matches a partial string (`always`, `first`, `rstmat`, etc).

alredy fixed with `$_WD_KEYS__MATCHTYPES`

**Comment by mlipok** [comment]
> > Also, would it be better to have this instead for the Func declaration?
> > `Func _WD_CapabilitiesAdd($key, $value1 = Default, $value2 = Default)`
> 
> As so far, my answer is: no, because ... look at my previous answer.

I will change it soon.

**Comment by mlipok** [comment]
> Needs to use __WD_Error throughout

Already done in recently merged PR

**Comment by mlipok** [comment]
> error handling should be checked
> 
> _Originally posted by @mlipok in [#263 (comment)](https://github.com/Danp2/au3WebDriver/pull/263#discussion_r830701133)_

SO this one is the last thing to do here.

**Comment by mlipok** [comment]
> > > Also, would it be better to have this instead for the Func declaration?
> > > `Func _WD_CapabilitiesAdd($key, $value1 = Default, $value2 = Default)`
> > 
> > 
> > As so far, my answer is: no, because ... look at my previous answer.
> 
> I will change it soon.

https://github.com/Danp2/au3WebDriver/pull/283/

**Comment by mlipok** [comment]
> > error handling should be checked
> > _Originally posted by @mlipok in [#263 (comment)](https://github.com/Danp2/au3WebDriver/pull/263#discussion_r830701133)_
> 
> So, as far as I see.... this one is the last thing to do here.

https://github.com/Danp2/au3WebDriver/pull/285

**Comment by Danp2** [comment]
> > Needs to use __WD_Error throughout
> 
> Already done in recently merged PR

One area where wd_capabilities.au3 differs from the rest of the UDF is on successful function return. Regular WD UDF functions do something like --

```autoit
    Return SetError(__WD_Error($sFuncName, $_WD_ERROR_Success, $sResponse), 0, $oResult)
```

P.S. I haven't yet looked at #290 to see if it affects any of the above

**Comment by mlipok** [comment]
> P.S. I haven't yet looked at #290 to see if it affects any of the above

So I will wait on the time you look on this.

**Comment by mlipok** [comment]
> One area where wd_capabilities.au3 differs from the rest of the UDF is on successful function return. Regular WD UDF functions do something like --
> 
> ```autoit
>     Return SetError(__WD_Error($sFuncName, $_WD_ERROR_Success, $sResponse), 0, $oResult)
> ```

Fixed
https://github.com/Danp2/au3WebDriver/pull/285#discussion_r850863286

**Comment by mlipok** [comment]
as this: https://github.com/Danp2/au3WebDriver/commit/e4d5f4f7004fca8b597d0d36987d12129b66aa7e
Was commited should we close this ISSUE ?
